### ISSUE TITLE
[question] - invalidate a result in cache when using "merge"

### ISSUE THREAD

**Issue by koukalp** [reduxjs/redux-toolkit#3733]
**Description**

I am trying to understand the right approach. Can you please give me an advice?

I am using a `query` enpoint with `merge` feature to "paginate" and keep all the results rendered on the page (clicking "next page" will display more results under those already displayed). 
I use the `merge` functionality to group all results together (along with `forceRefetch` and `serializeQueryArgs`). Which works. Now I want to be able to "remove" one of the results. Invalidating the cache (using `invalidateTags` or even calling `api.utils.invalidateTags` doesn't have the desired effect. While the query IS triggered, the cache is NOT updated and the deleted result is still displayed. 

In https://redux-toolkit.js.org/rtk-query/api/createApi#merge, I read: "no automatic structural sharing will be applied - it's up to you to update the cache appropriately"

What's the recommended way of achieving it? Reading the docs, I don't see a way to achieve that. Read some SO answers about using `entityAdapter` or using `prevPage`, `currentPage` and `nextPage`, but it just sounds too complicated approach for such a simple functionality I am trying to achieve = delete a single result from the cache, identified by it's ID

**My expectation**
(sort of what I thought would happen) I assumed the result would be invalidated the moment I define 
```
invalidateTags: ({id}) => [{type: "Result", id }] // just illustrates what I mean
```

Seems, when I opt for using `merge`, this feature gets skipped

**Comment by narasimhajupally** [comment]
facing the same problem, I want to delete the list of items I have merged till now when the tag is invalidated so that a fresh list is made

**Comment by narasimhajupally** [comment]
I am dealing with a recommendation endpoint (no query args) that gives a different set of results every time it is called so I used the merge function to append the results for every call but when I invalidate the tag for this endpoint the request is made and results are added to the existing list rather than replacing.

can the merge function have the information that the tag is invalidated? then we put a check there

**Comment by narasimhajupally** [comment]
another solution is to provide an option similar to `keepUnusedDataFor` like `deleteDataOnTagInvalidation` which if true deletes the current data and replaces it with API response.

**Comment by Placeblock** [comment]
Stumbled across the same problem today... Is there any solution for this?

**Comment by kylekirkby** [comment]
My team is also running into this issue right now.

Is the only option right now to manage the state independently from the RTK Query cache entries?

**Comment by markerikson** [comment]
I'm not sure I understand the problem that's being described. `merge` is a callback for updating the data inside of a given cache entry, by letting you combine the old data + the response data in some way.  It doesn't have anything to do with "invalidating a tag", which would trigger a refetch for a given entry.

Can folks give more details on what _exactly_ you're trying to accomplish?

**Comment by markerikson** [comment]
@haleyngonadi I'm not clear on what you expect to have happen, exactly.  

Are you expecting a full refetch to happen? Are you getting back data with a request, and the question is how to _replace_ the data that was already there using the `merge` callback?

**Comment by markerikson** [comment]
@haleyngonadi The `merge` callback is effectively a standard Immer `produce` function, which means that you can return a complete value instead of mutating the draft argument. So, you _should_ be able to literally `return responseData` inside of `merge` (instead of doing something like `cachedData.push(newStuff)`, and it _should_ replace the existing data in that cache entry.

**Comment by markerikson** [comment]
@haleyngonadi : yes, although there's two things that would need to change in that snippet:

- _return_ a value rather than mutating `cache`
- in the case of `createEntityAdapter` specifically, you'd need to do `return postAdapter.getInitialState()` (if the goal is to reset the data completely to empty), or `return postAdapter.setAll(postAdapter.getInitialState(), response)` (to _replace_ the data with the latest response contents)

**Comment by markerikson** [comment]
There's a couple tricky bits here:

- `merge` _only_ gets called when there is already data in the cache entry.  In other words, `merge` will _not_ be called for the _first_ successful request (the cache entry is empty, there isn't anything to "merge" into). It _will_ be called for _later_ successful responses (we already have data in the cache, there's more data, _now_ you can "merge" them).  This is logical, but also confusing.  
- That does mean that currently you would need to use the entity adapter in both `transformResponse` and in `merge`, because in the initial request case _only_ `transformResponse` will be called, and you still need the data to be normalized.

Going back to the question of "how do I _replace_ the existing data?", for the specific case you've got here, you'd want to use `postsAdapter.setAll(cache, response)`.  The issue is how does that logic in `merge` know if it _should_ do the "replace" vs the "update" handling.  I don't have an immediate answer for that, for multiple reasons: I don't know if we _do_ expose a way to know that into `merge` atm (ie, some argument that indicates "this was a `refetch()` response" or something); I don't know your specific codebase's needs; and also I'm replying to this while doing real day-job work and only have limited brainpower available atm :)

**Comment by koukalp** [comment]
@haleyngonadi - for me I need to delegate this to redux, where I can easily create reducers that either add ("onNextPageLoaded") items or remove a single item ("onDeleteItem") from the collection. 

It just feels like a missed opportunity, bc 
1) I am basically duplicating the rtk-query cache
2) I am not getting any benefits of the rtk-query cache while I add a lot of manual "labor" 

@markerikson - it feels the discussion went a bit off. My expectations are:

1. I use `rtk-query` with caching and `merge` function to implement infinite scroll to view a list of items
2. An item can be deleted from the list => in that case I want the deleted item to disappear from the cache. 
3. My issue is, `invalidateTags` stops to work the moment I use `merge` and I don't see a way to manipulate the cache to remove the (deleted) item

As mentioned to @haleyngonadi above, the solution with using redux and a few reducers is quite straightforward, however then I pretty much opt out the `rtk-query` cache which feels like a missed opportunity

**Comment by markerikson** [comment]
@koukalp yeah, there's a couple different sub-threads going on in this issue, which is also why I'm confused myself :)

My first note is that the whole "use `merge` as an infinite query implementation" thing is, frankly, a hack. And I was the one who came up with it, and I knew it was a hack at the time, and it's still a hack, and I don't like it.

I _want_ us to ship a "real" infinite query implementation at some point.  Ideally this year.  I've got another thread that's open to ask folks like you about specific use cases you would need it to handle, and I'd really appreciate it if you could leave some related thoughts over in that thread:

- https://github.com/reduxjs/redux-toolkit/discussions/3174

That said, it'll also be at least a couple months before we finally have time to comb through the list of open RTKQ feature requests, do some prioritization and planning, and _then_ actually start to work on them. So realistically I wouldn't expect us to be able to ship something like that until at least the end of the year, and the `merge` approach is what we've got to live with for now.

For _your_ specific case: I don't have enough context to understand what the actual technical issue is with "deleting an item" and "tag invalidation not working".  Could you provide an actual repo or CodeSandbox that shows this behavior?  What about tag invalidation is not working as expected?

**Comment by anujmv** [comment]
`import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";

// Define a service using a base URL and expected endpoints
export const pokemonApi = createApi({
  reducerPath: "pokemonApi",
  tagTypes: ["Pokemon"],
  baseQuery: fetchBaseQuery({ baseUrl: "https://pokeapi.co/api/v2/" }),
  endpoints: (builder) => ({
    getPokemonByName: builder.query({
      query: (page) => `pokemon?offset=${page * 20}&limit=20`,
      // Only have one cache entry because the arg always maps to one string
      serializeQueryArgs: ({ endpointName }) => {
        return endpointName;
      },
      // Always merge incoming data to the cache entry
      merge: (currentCache, newItems, { arg }) => {
        console.log(arg);
        if (arg === 0) {
          return newItems;
        }
        currentCache.results.push(...newItems.results);
      },

      // Refetch when the page arg changes
      forceRefetch({ currentArg, previousArg }) {
        return currentArg !== previousArg;
      },
      providesTags: ["Pokemon"],
    }),
  }),
});

// Export hooks for usage in functional components, which are
// auto-generated based on the defined endpoints
export const { useGetPokemonByNameQuery } = pokemonApi;
`

In react i manually set Page to 0 when needed to clear the cache

**Comment by luisfelipediaz** [comment]
I'm having the same issue, in my case, I also have the page size, a query filter, and other filters, so when I change any filter, I reset the current page to 1 and re-fetch the data, but I need to show the "skeleton" while the completely new filters are retrieving.

```javascript
export const usersServices = api.injectEndpoints({
  endpoints: build => ({
    getUsers: build.query({
      query: ({ companyId, page }) => ({
        url: EndPoints.USERS.GET_USERS,
        method: 'POST',
        body: buildQueryBody(companyId, page),
      }),
      providesTags: ['Users'],
      serializeQueryArgs: ({ endpointName, queryArgs }) => {
        return `${endpointName}("${queryArgs.companyId}", ${queryArgs.level}, "${queryArgs.searchQuery}")`;
      },
      forceRefetch({ currentArg, previousArg }, ...args) {
        if (!previousArg) return true;
        return (
          currentArg.companyId !== previousArg.companyId ||
          currentArg.page > previousArg.page ||
          currentArg.level !== previousArg.level ||
          currentArg.searchQuery !== previousArg.searchQuery
        );
      },
      merge: (currentData, responseData) => {
        currentData.pagination = responseData.pagination;
        currentData.users.push(...responseData.users);
      },
    }),
  }),
});
```

My workaround:

```javascript
const { data, isLoading, isFetching } = useGetUsersQuery({
    companyId,
    page,
    ...filters,
  });

// Here is my approach
const users = isFetching && page === 1 ? null : data;
```

So, when isFetching and is page 1 I'm setting users to null to simulate that the cache is null.

is there a way to prevent this validation?

**Comment by gimi-anders** [comment]
@markerikson I will give you a concrete example of when deleting an item will cause a problem with invalidation, when using infinite scroll.

Let's say there is 15 items in total on the server.
I fetch 10 items per page, and I now fetch page 1 and then fetch page 2, and merge the results from page 2 with the existing ones.
So the last query url called was: `https://myserver.com/api/users?page=2`

I now use a mutation endpoint to delete one item that was fetched on page 1. This mutation will invalidate the query above.
Problem is that the query above will now re-fetch page 2 since that was the last page fetched, and this page has not changed (or it has changed since it contains one less item, but the actual removed item was not affected).

It gets even worse if we started with 11 items on the server. In that case, when the deletion is done the invalidation will cause refetching of page 2 which in turn will return a 404 error from the server since that page does not even exist anymore.

**Comment by ievgen-klymenko-uvoteam** [comment]
Yes, we have almost the same issue. with `Infinite scroll`, as [gimi-anders ](https://github.com/reduxjs/redux-toolkit/issues/3733#issuecomment-1957209150).

`Automatic tag refetch` always calls the query with last arguments provided, which of course ends up in the same merge().

In our case, refetching all pages and erasing cache would be a solution. If `merge()`/`query()` calls would receive some indication about `'this particular call was a refetch'`, this would be possible to do I guess.

Manual cache updating is hard to implement here since it will need to be added in too many mutations.
For now, we will unfortunately have to invalidate some upper-tree (wrapping) queries to erase this cache completely.

But at least it works some way :) If @markerikson says it is was a `hack` solution from start, why would it work better than one :)

I wish @markerikson and team best of luck on this issue, since it sounds not an easy one.
Infinite scroll is used all around lately,
Can't wait to see what solution RTKQuery implements to fully support it!

**Comment by vancerbtw** [comment]
@haleyngonadi can you please provide your use of the query in the UI code? I am a bit confused by your use of 'otherArgs.arg?.last_seen_post_created_at' to determine if the merge is an invalidation that or a true merge. Trying to determine how to handle the merge has been my biggest struggle in regards to getting invalidateTags to work

**Comment by MrSachin7** [comment]
I am currently getting this issue as well.

So, the backend server paginates using the field 'created_at', so the newest items comes on the first page.
When I create a new item and invalidate the cache, it invalidates the last sent query , thus invalidating the query with the page number last sent.

If the last sent page number is not 1, the added item will not show until I hard refresh the page since, the added item is on page 1 on the backend which is never being refetched by RTK Query.

Does anyone have a quick workaround this ? I would even like a way to invalidate all queries regardless of the pageNumber.

**Comment by ievgen-klymenko-uvoteam** [comment]
I couldn't go with EntityAdapter solution that @haleyngonadi provided above, for reasons i cant remember.

My solution was straightforward, although not very intuitive.
But it required changes on backend (api query data results).

Here is api slice:
```
const { useGetPostsQuery } = query('getPosts', {
  query: ({ page }) =>
    API_POSTS_URL + (page > 1 ? `?page=${page}` : ''),

  serializeQueryArgs: ({ endpointName }) => endpointName,

  merge: (currentCache, newItems) => {
    const newPage = newItems.data?.page || 1
    if (currentCache.data.page === newPage) {
      return
    }

    const newPosts = newItems.data?.posts || []
    const hasMore = newItems.data?.hasMore || false

    currentCache.data.posts?.push(...newPosts)
    currentCache.data.page = newPage
    currentCache.data.hasMore = hasMore
  },

  forceRefetch: ({ currentArg, previousArg }) =>
    currentArg?.page !== previousArg?.page,
})
```

As you can see, we are manually comparing and saving `page` and `hasMore` "cursor meta info" right there, in the currentCache (where `posts` data is normally saved). And backend sends us this meta data along with every page query

Then, in the component, we control page number manually using usual state:
```
const Posts = () => {
  const [page, setPage] = useState(1)

  const { posts, hasMore, result } = useGetPostsQuery({ page })

  const loadMorePosts = () => setPage((prevPage) => prevPage + 1)

  return (
    <InfiniteScroll
      dataLength={posts.length}
      next={loadMorePosts}
      hasMore={hasMore}
    >
      {...posts}
    </InfiniteScroll>
  )
}
```

@MrSachin7 hope it helps!

**Comment by markerikson** [comment]
Circling back to this thread.

While this isn't a direct answer to the issues being discussed, we shipped full infinite query support in RTK 2.6.  That should remove the need to try to do it manually via `merge`.

For the specific case mentioned by @ensconced , our recommendation would be the same as what React Query suggests.  Manually update the cache of the infinite query endpoint to reset the `{pages, pageParams}` arrays so they're empty, then have the refetch happen.

I'm going to go ahead and close this issue. It's probable there's something that might not have been resolved - please let me know if that's the case!
